### PR TITLE
Add SearchContainer and ISearchable

### PR DESCRIPTION
**Pull request by Jorolf**

### CODE REVIEW

File: osu.Framework/Graphics/Containers/ISearchable.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using System;
+using System.Collections.Generic;
+
+namespace osu.Framework.Graphics.Containers
+{
+    public interface ISearchable
+    {
+        string[] Keywords { get; }
+
+        bool LastMatch { get; set; }
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ISearchable.cs]: this should not be here

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                        if (AreMatching(keyword, Filter))
+                            contains = true;
+
+                    foreach (string keyword in parentKeywords)
+                        if (AreMatching(keyword, Filter))
+                            contains = true;
+
+                    search.LastMatch = contains;
+                    if (contains)
+                        OnMatch(drawable);
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: the drawable itself should probably handle the onmatch/mismatch stuff. via a single bool in the interface.

File: osu.Framework.VisualTests/Tests/TestCaseSearchContainer.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        private class SearchableText : SpriteText, ISearchable
+        {
+            public string[] Keywords => new [] { Text };
+            public bool LastMatch { get; set; }
+        }
+
+        private class SearchableFlowContainer<T> : FillFlowContainer<T>, ISearchableChildren where T : Drawable
+        {
+            public string[] Keywords => new[] { "flowcontainer" };
```
Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseSearchContainer.cs]: this is very misleading in the testcase

File: osu.Framework.VisualTests/Tests/TestCaseSearchContainer.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        public override void Reset()
+        {
+            base.Reset();
+
+            text = new List<SearchableText>();
+            for(int i = 0; i < 8; i++)
+            {
+                text.Add(new SearchableText
+                {
+                    Text = RNG.Next(1000).ToString(),
```
Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseSearchContainer.cs]: test cases shouldn't be random

File: osu.Framework.VisualTests/Tests/TestCaseSearchContainer.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        private SearchContainer search;
+        private TextBox textBox;
+        private List<SearchableText> text;
+
+        public override void Reset()
+        {
+            base.Reset();
+
+            text = new List<SearchableText>();
```
Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseSearchContainer.cs]: bad variable name

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+namespace osu.Framework.Graphics.Containers
+{
+    public class SearchContainer : SearchContainer<Drawable>, ISearchableChildren
+    { }
+
+    public class SearchContainer<T> : Container<T> where T : Drawable
+    {
+        private string filter;
+        public string Filter
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: why not SearchTerm or something related?

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                });
+            }
+        }
+
+        public IEnumerable<Drawable> SearchableContent { get; set; } = null;
+
+        public string[] Keywords => null;
+        public bool LastMatch { get; set; }
+        public IEnumerable<Drawable> SearchableChildren => SearchableContent ?? Children;
+        public Action AfterSearch { get; set; }
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: xmldoc comments for everything.

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+        public int RematchDelay { get; set; } = 200;
+
+        public delegate void OnSearchHandler(Drawable searchable);
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: why not Action

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+        public OnSearchHandler OnMatch { get; set; } = delegate { };
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: why are you initialistinbg with a delegate of nothing

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        public OnSearchHandler OnMismatch { get; set; } = delegate { };
+
+        protected virtual bool AreMatching(string value1, string value2) => value1?.IndexOf(value2, StringComparison.InvariantCultureIgnoreCase) >= 0;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: is this ever going to be overridden

**Comment by peppy** [comment]
There's so many issues with this PR i honestly don't know where to start. The code style is very hard to follow and not up to our standards. There are redundant calls and fields in the wrong place that need to be thought out again from the start.

I'm going to close this for now. Feel free to take another attempt at it with the following in mind:
- A high level object should not decide how to transition a child (or a deeply nested child via recursion as it may be in this case).
- If you have a boolean state, use a single action or property to handle it, not two separate void delegates.
- Comment your public fields and organise them in a way that doesn't hurt the brain.

File: osu.Framework/Graphics/Containers/SearchContainer.cs (hunk 10 above)
Comment (Jorolf) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: This should have been used for song select so custom search terms could be implemented (cs<4, ar=4.7, etc.)
